File: docker/web/webrtc_video/src/signaling.py
import asyncio
import websockets
import json

# Store connected clients
clients = set()

SERVER_IP = "0.0.0.0"
SERVER_PORT = 8765

async def signaling_handler(websocket):
    clients.add(websocket)
    print("Client connected")
    try:
        async for message in websocket:
            data = json.loads(message)
            print(f"Received: {data}")

            # Forward messages to all other clients
            for client in clients:
                if client != websocket:
                    await client.send(json.dumps(data))

    except websockets.exceptions.ConnectionClosed:
        print("Client disconnected")
    finally:
        clients.remove(websocket)

async def main():
    server = await websockets.serve(signaling_handler, SERVER_IP, SERVER_PORT)
    print(f"Signaling server running on ws://{SERVER_IP}:{SERVER_PORT}")
    await server.wait_closed()

asyncio.run(main())

File: docker/web/webrtc_video/src/signaling.cpp
#include <websocketpp/config/asio_no_tls.hpp>
#include <websocketpp/server.hpp>
#include <set>
#include <string>
#include <jsoncpp/json/json.h>

typedef websocketpp::server<websocketpp::config::asio> server;

class signaling_server {
public:
    signaling_server() {
        m_server.init_asio();
        m_server.set_open_handler(std::bind(&signaling_server::on_open, this, std::placeholders::_1));
        m_server.set_close_handler(std::bind(&signaling_server::on_close, this, std::placeholders::_1));
        m_server.set_message_handler(std::bind(&signaling_server::on_message, this, std::placeholders::_1, std::placeholders::_2));
    }

    void run(uint16_t port) {
        std::cout << "Starting signaling server on port " << port << std::endl;
        m_server.listen(port);
        m_server.start_accept();
        m_server.run();
    }

private:
    void on_open(websocketpp::connection_hdl hdl) {
        m_connections.insert(hdl);
        std::cout << "Client connected" << std::endl;
    }

    void on_close(websocketpp::connection_hdl hdl) {
        m_connections.erase(hdl);
        std::cout << "Client disconnected" << std::endl;
    }

    void on_message(websocketpp::connection_hdl hdl, server::message_ptr msg) {
        // Parse incoming JSON message
        std::string payload = msg->get_payload();
        Json::Value root;
        Json::Reader reader;

        if (!reader.parse(payload, root)) {
            std::cerr << "Failed to parse JSON message: " << reader.getFormattedErrorMessages() << std::endl;
            return;
        }

        // Ensure we're sending a text message (opcode 0x1)
        if (msg->get_opcode() == websocketpp::frame::opcode::text) {
            // Forward the message to all other clients
            for (auto it : m_connections) {
                if (it.lock() != hdl.lock()) {
                    m_server.send(it, payload, msg->get_opcode());
                }
            }
        }
    }
    server m_server;
    std::set<websocketpp::connection_hdl, std::owner_less<websocketpp::connection_hdl>> m_connections;
};

int main() {
    signaling_server server_instance;
    server_instance.run(8765); // Run server
    return 0;
}